### DISCUSSION TITLE
Unknown number of custom pages that are read from database

### DISCUSSION THREAD

**Discussion by KarlisJ** [filamentphp/filament#6865] [Help] [answered]
Here are our requirements:
- We have a `pages` table with corresponding `Pages` model
- A `page` has `title` attribute and additional information that we use to display information in the page
- Each user/organization might have different number of pages with different content

We want to read `pages`, add them to the navigation, and use a single page template to display information. 

How can we accomplish that with Filament? Filament seems to expect to have a dedicated file for each custom page.

**Comment by danharrin** [comment]
Look, what you're trying to achieve is very custom and you should try and work out how to do it yourself first, we can't really write such a large system for you. The trick is to register a route with a wildcard parameter and then decide which page to render based on it. Navigation items can be registered from a service provider.

**Comment by KarlisJ** [comment, marked as answer]
I created a working PoC and will definitely polish it but sharing here as an answer to my question for others to use.

First, create custom page and disable automatic navigation registration
```php
protected static bool $shouldRegisterNavigation = false;
```

and override `getRoutes` method
```php
public static function getRoutes(): Closure
{
    return function () {
        Route::get('/custom/{variable}', static::class)
            ->middleware(static::getMiddlewares())
            ->name('variable');
    };
}
```
- I'm keeping the `->middleware(static::getMiddlewares())` as Filament relies on that when registering routes. 
- Prefixing the route with `/custom` is intentional. Filament registers routes for Pages first and then Resources. So without a unique prefix, the route will match any resources first.

Also, getting the `variable` from route to use in the page when creating the page
```php
public function mount( $variable )
{
    // Do what you want with route parameter
}
```

Lastly, I need to register navigation items in service provider
```php
public function boot()
{
    Filament::serving(function () {
        $pages = Pages::all();
        foreach ($pages as $page ) {
            Filament::registerNavigationItems([
                NavigationItem::make()
                    ->label($page->label)
                    ->icon($page->icon)
                    ->url(route('filament.pages.variable', ['variable' => $page->url_slug]))
                    // ...
            ]);
        }
    });
}
```
